### PR TITLE
obs-qsv11: Add AV1 changes

### PR DESCRIPTION
**Pull request by thyintel**
### Description

- Add Intel AV1 hardware encoder

### Motivation and Context

- We should have a way to use Intel AV1 hardware encoder in streaming and recording.

### How Has This Been Tested?

- Integrated graphics (to ensure the original code path works)
- Intel CPU and Intel Arc GPU
- Non-Intel and Intel Arc GPU

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by pkviet** [comment]
thanks for this PR, which we're quite happy to see. Could you drop the merge commit by doing a proper rebase to master branch of obs ?

File: plugins/obs-qsv11/QSV_Encoder_Internal.cpp [hunk 1]
```diff
@@ -333,6 +367,64 @@ bool QSV_Encoder_Internal::InitParams(qsv_param_t *pParams)
...
+		m_mfxEncParams.mfx.BufferSizeInKB = pParams->nTargetBitRate * 2;
+		m_mfxEncParams.mfx.InitialDelayInKB =
+			pParams->nTargetBitRate * 1;
+		break;
+	default:
+		break;
+	}
+
+	// Width must be a multiple of 16
+	// Height must be a multiple of 16 in case of frame picture and a
```
Comment (derrod) [inline plugins/obs-qsv11/QSV_Encoder_Internal.cpp:413]: Does this still apply here for AV1? And will the output have the correct resolution regardless?

**Comment by derrod** [comment]
Also I assume our media sdk version needs to be updated?

But even the one available at https://github.com/Intel-Media-SDK/MediaSDK does not appear to support AV1 encoding (yet)?

**Comment by pkviet** [comment]
ps: git rebase -i origin/master
Then in the editor window, drop the merge commit.
Force push after that.

**Comment by Kieaer** [comment]
It's work on Intel A380
currently support CBR and VBR

**Comment by A1RM4X** [comment]
Tested with an Intel A380 on Windows.

Works with CBR and VBR.

**Comment by RavenX8** [comment]
Tested with Intel A750 on Windows 10 Pro 21H2
CBR and VBR working

**Comment by ma3uk** [comment]
Can anyone explain how it works without the OneVPL library?

File: plugins/obs-qsv11/QSV_Encoder_Internal.cpp [hunk 2]
```diff
@@ -188,6 +221,7 @@ mfxStatus QSV_Encoder_Internal::Open(qsv_param_t *pParams)
 	if (sts >= MFX_ERR_NONE) {
 		g_numEncodersOpen++;
 	}
+
```
Comment (RytoEX) [inline plugins/obs-qsv11/QSV_Encoder_Internal.cpp:224]: Avoid unrelated whitespace changes.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Some nits. Please also examine our changes in #6679 and see if they apply here. We had issues with MSDK seemingly returning incorrect timestamps for fractional FPS.

**Comment by NyaakoYT** [comment]
Tested with Intel A380 on Windows 11 Home 22H2 Build 22621.755

CBR and VBR working

**Comment by RytoEX** [comment]
Thank you for all of the comments confirming that this works on your systems. At this time, we do not need more confirmations, and we'd like to reserve this thread for discussion of the code and implementation itself. Thank you for your understanding.

File: plugins/obs-qsv11/QSV_Encoder.cpp [hunk 3]
```diff
@@ -144,27 +147,40 @@ bool prefer_current_or_igpu_enc(int *iGPUIndex)
...
+{
+	bool av1_supported = false;
+
+	QSV_Encoder_Internal *pEncoder = new QSV_Encoder_Internal(impl, ver);
+
+	av1_supported = pEncoder->AV1_Supported(impl);
+
+	delete pEncoder;
+	if (pEncoder)
+		is_active.store(false);
```
Comment (derrod) [inline plugins/obs-qsv11/QSV_Encoder.cpp:181]: What exactly is the function of this `delete <x>; if(x)` pattern? Also `is_active` doesn't seem to ever be read?

File: plugins/obs-qsv11/QSV_Encoder.cpp [hunk 4]
```diff
@@ -175,12 +191,22 @@ void qsv_encoder_version(unsigned short *major, unsigned short *minor)
...
-	if (prefer_current_or_igpu_enc(&igpu_index) &&
-	    (igpu_index < _countof(impl_list))) {
-		impl = impl_list[igpu_index];
+
+	//select current adapter - will be iGPU if exists due to adapter reordering
+	impl = impl_list[adapter_info->currentGPUIndex];
+
+	//check iGPU support for AV1 first, if dGPU exists check support for AV1
+	if (pParams->codec == MFX_CODEC_AV1) {
+		bool av1_supported = qsv_av1_supported(impl);
```
Comment (derrod) [inline plugins/obs-qsv11/QSV_Encoder.cpp:206]: This appears to check if AV1 encoding is supported every time an encoder is started, I assume due to the possibility that the user has selected a GPU that doesn't support it.

Since this check already happens in `obs_module_load`, would it perhaps be possible to cache this information and simply look it up here instead?

**Review by derrod — COMMENTED** [review]
Also to clarify; As far as I can tell this PR does not add support for HDR metadata or input formats?

File: plugins/obs-qsv11/QSV_Encoder.cpp (hunk 4 above)
Comment (derrod) [reply plugins/obs-qsv11/QSV_Encoder.cpp:206]: As a follow-up, is there perhaps a more graceful way to check capabilities than creating an encoder?

File: plugins/obs-qsv11/QSV_Encoder_Internal.cpp (hunk 1 above)
Comment (thyintel) [reply plugins/obs-qsv11/QSV_Encoder_Internal.cpp:412]: This does apply for AV1. Any resolution given will be correct as long as the MSDK_ALIGN16 is used.

**Comment by thyintel** [comment]
> Also I assume our media sdk version needs to be updated?
> 
> But even the one available at https://github.com/Intel-Media-SDK/MediaSDK does not appear to support AV1 encoding (yet)?

MSDK 1.35 API added support for AV1 to allow users more time to transition to OneVPL. Unfortunately, the documentation  for MSDK is lagging. Though if you look at mfxstructures.h you can see the AV1 features that are supported through MSDK. Any new features which don’t exist in the MSDK headers will be added only to OneVPL.

**Comment by Kieaer** [comment]
There's no AV1 Streaming.

**Comment by GhostofReachCashGrab** [comment]
Tested and found using mov crashes 
image
```
22:09:52.869: >>> new qsv encoder
22:09:52.870: [qsv encoder: 'advanced_video_recording'] settings:
22:09:52.870: 	rate_control:   CBR
22:09:52.870: 	fps_num:        60
22:09:52.870: 	fps_den:        1
22:09:52.870: 	width:          2560
22:09:52.870: 	height:         1440
22:09:52.870: [qsv encoder: 'advanced_video_recording'] debug info:
22:09:52.877: 	impl:           D3D11
22:09:52.877: 	surf:           D3D11
22:09:53.998: 	impl:           D3D11
22:09:53.998: 	surf:           D3D11
22:09:55.106: 	impl:           D3D11
22:09:55.106: 	surf:           D3D11
22:09:55.211: 	m_nSurfNum:     9
22:09:55.223: 	m_nTaskPool:    4
22:09:55.223: 	major:          1
22:09:55.223: 	minor:          255
22:09:55.223: ---------------------------------
22:09:55.223: [FFmpeg aac encoder: 'Track1'] bitrate: 320, channels: 2, channel_layout: 3
22:09:55.223: 
22:09:55.224: ---------------------------------
22:09:55.224: [FFmpeg aac encoder: 'Track2'] bitrate: 320, channels: 2, channel_layout: 3
22:09:55.224: 
22:09:55.234: ==== Recording Start ===============================================
22:09:55.234: [ffmpeg muxer: 'adv_file_output'] Writing file 'C:/Users/ghost/Videos/2022-11-09 22-09-51.mov'...
22:09:55.267: [qsv encoder: 'advanced_video_recording'] settings:
22:09:55.267: 	rate_control:   CBR
22:09:55.267: 	fps_num:        60
22:09:55.267: 	fps_den:        1
22:09:55.267: 	width:          2560
22:09:55.267: 	height:         1440
22:09:55.267: [qsv encoder: 'advanced_video_recording'] debug info:
22:09:55.399: 	m_nSurfNum:     9
22:09:55.413: 	m_nTaskPool:    4
22:09:56.443: [ffmpeg muxer: 'adv_file_output'] os_process_pipe_write for packet data failed
22:09:56.443: [ffmpeg muxer: 'adv_file_output'] ffmpeg-mux: Error opening 'C:/Users/ghost/Videos/2022-11-09 22-09-51.mov': Invalid argument
22:09:56.444: [ffmpeg muxer: 'adv_file_output'] Output of file 'C:/Users/ghost/Videos/2022-11-09 22-09-51.mov' stopped
22:09:56.444: Output 'adv_file_output': stopping
22:09:56.444: Output 'adv_file_output': Total frames output: 1
22:09:56.444: Output 'adv_file_output': Total drawn frames: 73
22:09:56.445: ==== Recording Stop ================================================
22:09:56.489: warning: 2 frames left in the queue on closing
22:09:56.489: warning: 2 frames left in the queue on closing
```

**Comment by derrod** [comment]
Yes, the mov container doesn't support AV1 (yet) so that will error out. Use MKV or MP4.

**Comment by RyzenDew** [comment]
> Yes, the mov container doesn't support AV1 (yet) so that will error out. Use MKV or MP4.

Mkv + av1 is not supported by my video editor so that sucks

File: plugins/obs-qsv11/QSV_Encoder.cpp (hunk 4 above)
Comment (thyintel) [reply plugins/obs-qsv11/QSV_Encoder.cpp]: We have cleaned up this portion of the code. Please take a look and let us know any other feedback.

Comment (derrod) [reply plugins/obs-qsv11/QSV_Encoder.cpp]: Yeah this seems good, I don't have the hardware to check functionality so that'll be up to @RytoEX 

There's still the question of what happens with multiple dGPUs, but I'm fine leaving that for later since those setups are exceedingly rare anyway.

Comment (RytoEX) [reply plugins/obs-qsv11/QSV_Encoder.cpp]: Will re-test as soon as I can.

Comment (NyaakoYT) [reply plugins/obs-qsv11/QSV_Encoder.cpp]: > Yeah this seems good, I don't have the hardware to check functionality so that'll be up to @RytoEX
> 
> There's still the question of what happens with multiple dGPUs, but I'm fine leaving that for later since those setups are exceedingly rare anyway.

I'm using two GPUs on a single PC and the QSV encoder doesn't start up if OBS is running on a non-Intel GPU (like NVIDIA/AMD). It can be done on the stable OBS build QSV encoder, however.

Comment (RyzenDew) [reply plugins/obs-qsv11/QSV_Encoder.cpp]: Having the ability to use av1 with another vendor gpu in the system is a must it’s why these gpus are being bought so the question is can this be done?

**Comment by RyzenDew** [comment]
After loads of testing it's been very stable.
no crashing
no weird bugs
no corrupt footage 
works well with xenia which is a huge thing while gaming.
also shared the PR in the intel arc gpu discord they also did testing including intel devs and no issues yet

File: plugins/obs-qsv11/obs-qsv11.c [hunk 5]
```diff
@@ -566,13 +624,102 @@ static void update_params(struct obs_qsv *obsqsv, obs_data_t *settings)
...
+	/* internal convenience parameter, overrides rate control param
+	 * XXX: Deprecated */
+	if (cbr_override) {
+		warn("\"cbr\" setting has been deprecated for all encoders!  "
+		     "Please set \"rate_control\" to \"CBR\" instead.  "
+		     "Forcing CBR mode.  "
+		     "(Note to all: this is why you shouldn't use strings for "
+		     "common settings)");
+		rate_control = "CBR";
+	}
```
Comment (RytoEX) [inline plugins/obs-qsv11/obs-qsv11.c:679]: This was an override for a setting that was replaced over seven years ago. I don't think we need to carry this over to a brand new encoder, since `settings` are encoder specific.

File: plugins/obs-qsv11/obs-qsv11.c [hunk 6]
```diff
@@ -566,13 +624,102 @@ static void update_params(struct obs_qsv *obsqsv, obs_data_t *settings)
...
+
+	const char *target_usage =
+		obs_data_get_string(settings, "target_usage");
+	const char *profile = obs_data_get_string(settings, "profile");
+	const char *rate_control =
+		obs_data_get_string(settings, "rate_control");
+	int target_bitrate = (int)obs_data_get_int(settings, "bitrate");
+	int max_bitrate = (int)obs_data_get_int(settings, "max_bitrate");
+	int keyint_sec = (int)obs_data_get_int(settings, "keyint_sec");
+	bool cbr_override = obs_data_get_bool(settings, "cbr");
```
Comment (RytoEX) [inline plugins/obs-qsv11/obs-qsv11.c:640]: This was an override for a setting that was replaced over seven years ago. I don't think we need to carry this over to a brand new encoder, since `settings` are encoder specific.

File: plugins/obs-qsv11/obs-qsv11.c [hunk 7]
```diff
@@ -106,13 +108,35 @@ static unsigned short g_verMinor;
...
+}
+
+bool av1_supported(mfxIMPL impl)
+{
+	return qsv_av1_supported(impl);
+}
+
+bool check_av1_adapter()
+{
+	//av1_soft is non texture based encode path
```
Comment (RytoEX) [inline plugins/obs-qsv11/obs-qsv11.c:132]: Leave a space between the `//` and the actual comment.
Use a hyphen for "non-texture".

File: plugins/obs-qsv11/QSV_Encoder_Internal.h [hunk 8]
```diff
@@ -63,7 +63,6 @@ class QSV_Encoder_Internal {
 public:
 	QSV_Encoder_Internal(mfxIMPL &impl, mfxVersion &version);
 	~QSV_Encoder_Internal();
-
```
Comment (RytoEX) [inline plugins/obs-qsv11/QSV_Encoder_Internal.h:66]: Undo this removal.

**Review by RytoEX — CHANGES_REQUESTED** [review]

**Comment by NyaakoYT** [comment]
Intel ARC AV1 working well on the latest build (g654c60caf). 
The lag gap to start recording has shortened from what I can tell, so that's nice.
Still can't use the QSV encoder (h264/av1) if OBS is running on a separate GPU (like a NVIDIA GPU). Could do it with the QSV encoder (h264) on OBS 28.1.2 stable, so there's probably still some things to get done.
Looking forward for when that time comes.

File: plugins/obs-qsv11/obs-qsv11.c [hunk 9]
```diff
@@ -962,6 +1162,71 @@ static void parse_packet(struct obs_qsv *obsqsv, struct encoder_packet *packet,
...
+		if (g_bFirst) {
+			packet->dts = packet->pts - 3 * obsqsv->params.nFpsDen;
+		} else if (pFrame) {
+			packet->dts = packet->pts - 10 * obsqsv->params.nFpsDen;
+			g_prevDts = packet->dts;
+		} else {
+			packet->dts = g_prevDts + obsqsv->params.nFpsDen;
+			g_prevDts = packet->dts;
+		}
+	} else {
```
[the remark this answers is not in the document]
Comment (thyintel) [reply plugins/obs-qsv11/obs-qsv11.c:1200]: Not yet. We’re actually refining automatic DTS function in terms of accuracy. Please use existing manual solution until we complete it.

[the remark this answers is not in the document]
Comment (RytoEX) [reply plugins/obs-qsv11/obs-qsv11.c:1200]: What I'm saying is that `g_pts2dtsShift` is defined here:
* https://github.com/obsproject/obs-studio/blob/28.1.2/plugins/obs-qsv11/obs-qsv11.c#L669-L686

Which suggests that this will never actually use the manual calculation in the AV1 branch, because AV1 is not supported in MSDK 1.6 or lower.

[the remark this answers is not in the document]
Comment (thyintel) [reply plugins/obs-qsv11/obs-qsv11.c]: Sorry for the misunderstanding! You are correct and we had a copy-paste error on our side. It should be fixed now.

File: plugins/obs-qsv11/QSV_Encoder.cpp [hunk 10]
```diff
@@ -144,27 +149,71 @@ bool prefer_current_or_igpu_enc(int *iGPUIndex)
...
+				adapter_info->hasIGPU = true;
+				adapter_info->iGPUIndex = adapterIndex;
+				adapter_info->iGPUAV1 =
+					qsv_av1_supported(MFX_IMPL_HARDWARE2);
 			} else {
-				hasDGPU = true;
+				adapter_info->hasDGPU = true;
+				adapter_info->dGPUIndex = adapterIndex;
+				adapter_info->dGPUAV1 =
+					qsv_av1_supported(MFX_IMPL_HARDWARE);
```
Comment (jp9000) [inline plugins/obs-qsv11/QSV_Encoder.cpp:166]: Respectfully, this code upsets me a lot. You have to initialize an AV1 encoder just to use an H264 encoder. This adds many seconds of initialization time to not just the AV1 encoder, but sadly, to the H264 encoder as well.

This is a terrible situation that no one should find acceptable.

Comment (jp9000) [reply plugins/obs-qsv11/QSV_Encoder.cpp:166]: I'm going to merge the AV1 encoder without this diff for beta. The people over there at Intel have to figure out a better way to handle this situation. There's about 5 weeks for testing before final release. Something else needs to be figured out before then.

I will not accept this particular change that requires spawning multiple encoder sessions per singular encode initialization.

Comment (jp9000) [reply plugins/obs-qsv11/QSV_Encoder.cpp:166]: I'm sorry, I just have a lot of stress lately and I didn't mean to be like this. In the future though, just remember to keep in mind initialization times.

The way this should be handled is by querying support on startup. If it's unavoidable to test codec support except by spawning and encoder, then it should probably be done on startup, not when initializing an encoder.

Again, I apologize for my harshness. I am just under a lot of stress right now and you don't deserve that.
